### PR TITLE
Add The Boehm-Demers-Weiser conservative Garbage Collector

### PR DESCRIPTION
**Pull request by PunchCakee**
This PR resolves issue #2338 by adding the GC to xmake-repo.

### CODE REVIEW

File: packages/b/bdwgc/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package("bdwgc")
+    set_homepage("https://www.hboehm.info/gc/")
+    set_description("The Boehm-Demers-Weiser conservative C/C++ Garbage Collector (bdwgc, also known as bdw-gc, boehm-gc, libgc)")
+
+    add_urls("https://github.com/ivmai/bdwgc/-/archive/$(version).tar.gz",
+             "https://github.com/ivmai/bdwgc.git")
+
+    add_versions("v8.2.4", "18e63ab1428bd52e691da107a6a56651c161210b11fbe22e2aa3c31f7fa00ca5")
+    add_deps("cmake")
```
Comment (waruqi) [inline packages/b/bdwgc/xmake.lua:9]: please add an empty line

File: packages/b/bdwgc/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package("bdwgc")
+    set_homepage("https://www.hboehm.info/gc/")
+    set_description("The Boehm-Demers-Weiser conservative C/C++ Garbage Collector (bdwgc, also known as bdw-gc, boehm-gc, libgc)")
+
+    add_urls("https://github.com/ivmai/bdwgc/-/archive/$(version).tar.gz",
+             "https://github.com/ivmai/bdwgc.git")
+
+    add_versions("v8.2.4", "18e63ab1428bd52e691da107a6a56651c161210b11fbe22e2aa3c31f7fa00ca5")
```
Comment (waruqi) [inline packages/b/bdwgc/xmake.lua:8]: add empty line

File: packages/b/bdwgc/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    on_test(function (package)
+        assert(package:has_cfuncs({test=[[
+        
+        int main()
+        {
+            GC_INIT();
+        
+            int *ptr = GC_MALLOC(sizeof(int));
+            *ptr = 42;
+
```
Comment (waruqi) [inline packages/b/bdwgc/xmake.lua:26]: please remove empty line

File: packages/b/bdwgc/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        
+            int *ptr = GC_MALLOC(sizeof(int));
+            *ptr = 42;
+        
+            printf("Value: %d\n", *ptr);
+        
+            return 0;
+        }
+
+        ]]}),{configs = {includes = "gc"}})
```
Comment (waruqi) [inline packages/b/bdwgc/xmake.lua:32]: `include <gc>`? no `.h` suffix?

File: packages/b/bdwgc/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    on_install(function (package)
+        local configs = {}
+        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        import("package.tools.cmake").install(package, configs)
+    end)
+
+    on_test(function (package)
+        assert(package:has_cfuncs({test=[[
+
```
Comment (waruqi) [inline packages/b/bdwgc/xmake.lua:19]: remove empty line

File: packages/b/bdwgc/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        local configs = {}
+        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        import("package.tools.cmake").install(package, configs)
+    end)
+
+    on_test(function (package)
+        assert(package:has_cfuncs({test=[[
+        
+        int main()
```
Comment (waruqi) [inline packages/b/bdwgc/xmake.lua:20]: use `void test()` instead of main

File: packages/b/bdwgc/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    add_deps("cmake")
+    on_install(function (package)
+        local configs = {}
+        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        import("package.tools.cmake").install(package, configs)
+    end)
+    on_test(function (package)
+        assert(package:has_cfuncs({test=[[
+        void main()
```
Comment (waruqi) [inline packages/b/bdwgc/xmake.lua:19]: do not use main, rename to `test()`

Comment (PunchCakee) [reply packages/b/bdwgc/xmake.lua]: fixed

**Comment by PunchCakee** [comment]
> if you want to support mingw/windows, you can open a new pr and improve it.

What do you mean? It works on all platforms according to the CI, i have tested it on mingw too.

**Comment by waruqi** [comment]
> > if you want to support mingw/windows, you can open a new pr and improve it.
> 
> What do you mean? It works on all platforms according to the CI, i have tested it on mingw too.

it fails on mingw/windows ci, so I disabled them.

**Comment by waruqi** [comment]
mingw: https://github.com/xmake-io/xmake-repo/actions/runs/5582045357/jobs/10200901958
windows: https://github.com/xmake-io/xmake-repo/actions/runs/5582045373/jobs/10200903358

**Pull request merged**